### PR TITLE
PowerDNS Forward-Notify Patch

### PR DESCRIPTION
**Pull request by DrRemorse**
This patch will allow you to redirect PowerDNS notifications to a "passthru" server. It is intended to fix notify/axfr behavior in anycast clusters of PowerDNS auth servers however it may be useful for other situations.

The configuration option "passthru-notify" has been added to the pdns.conf parser. The option accepts multiple IPv4 and IPv6 address values.

This patch obsoletes issue #655.

Ralph Covelli
Hurricane Electric / AS6939
rcovelli@he.net

### CODE REVIEW

**Comment by Habbie** [comment]
Thank you for this patch!

I'm happy to merge this with some changes:
1. a rebase (we changed a lot of code on our end, sorry about that)
2. making sure Travis passes
3. a rename to `forward-notify` or something similar
4. the `testIPv6addr` function (and the v4 equivalent) almost certainly duplicate code we already have, can you look into cleaning that up? I presume ComboAddress can already do the checking.
5. clear documentation; right now this feature both forwards incoming NOTIFYs and overrides the target of locally initiated NOTIFYs. Given the override can also be done with a combination of `also-notify` and `only-notify` I am somewhat tempted to remove the override feature in full.

Then, we need a review to make sure the `static Resolver passthruResolver` really is safe, I'm happy to do that.

We're happy to help with any of the above steps if you have trouble with them or can't find the time. Do let us know!

**Comment by klaus3000** [comment]
Which problem are you actually trying to solve? There are plenty of notification options already in PDNS.

**Comment by klaus3000** [comment]
Hi DrRemorse! I still do not understand your problem. We also use PDNS in Anycast Clusters but do not have problems. There are plenty of options like notify-only, slave-renotify, ALSO-NOTIFY and what we do, we changed some default SQL queries to a SQL function which has some additional notify logic, so we do lots of notify tricks without any code changes.

**Comment by DrRemorse** [comment]
Hi,

We operate a large anycast cluster of PowerDNS servers as well. Our servers are not configured as masters or slaves.  Internally the backend handles synchronization however we can also be configured to slave zones off other user's servers.  In this case when the user's server sends us a notification we use this feature to relay that notification to an AXFR controller which then in turn pulls the zone from the user's server.  There may be other methods of accomplishing this however we found this way efficient and scalable.  We are offering this to the community in the hopes that someone will find it as useful as we did.

Thanks!

**Comment by klaus3000** [comment]
Basically our setup may be similar: 1 "control server" PDNS fetching the zone from the customers and receiving NOTIFYs from the customers. Zone are configured with type=SLAVE. The backend is replicated to the anycast nodes (Postgresql+Slony) and the anycast nodes operate with type=NATIVE. To prevent the "control server" to send automatic NOTIFYs to the NS hostnames we use "only-notify=". Customers can also fetch their zones from our control server. Therefore we have slave-renotify=yes and set ALSO-NOTIFY for these customer's zones. So, no code big changes are needed (only one optimizations: https://github.com/PowerDNS/pdns/pull/4861)

File: pdns/communicator.cc [hunk 1]
```diff
@@ -92,6 +92,19 @@ void CommunicatorClass::go()
...
+  vector<string> forwards;
+  stringtok(forwards, ::arg()["forward-notify"], ", \t");
+  for (vector<string>::const_iterator iter = forwards.begin(); iter != forwards.end(); ++iter) {
+    try {
+      ComboAddress caIp(*iter, 53);
+      PacketHandler::s_forwardNotify.insert(caIp.toStringWithPort());
+    }
+    catch(PDNSException &e) {
+      L<<Logger::Error<<"Unparseable IP in forward-notify. Error: "<<e.reason<<endl;
+      exit(1);
```
Comment (pieterlexis) [inline pdns/communicator.cc]: please don't kill the daemon on error

Comment (DrRemorse) [inline pdns/communicator.cc]: Hi,

It was written like this to be consistent with the behavior of also-notify, only-notify and friends which directly precede this chunk of code.  This code is only run once at pdns startup.

Thanks!

**Comment by Habbie** [comment]
After IRC discussion: needs some squashing, but given the small size of the total change, please feel free to squash it all into one commit.

File: pdns/packethandler.cc [hunk 2]
```diff
@@ -886,7 +888,17 @@ int PacketHandler::processNotify(DNSPacket *p)
     
   // ok, we've done our checks
   di.backend = 0;
-  Communicator.addSlaveCheckRequest(di, p->d_remote);
+
+  if(!::arg()["forward-notify"].empty()) {
```
Comment (Habbie) [inline pdns/packethandler.cc]: It's cheaper to check if `PacketHandler::s_forwardNotify` is empty.

File: pdns/packethandler.cc [hunk 3]
```diff
@@ -886,7 +888,17 @@ int PacketHandler::processNotify(DNSPacket *p)
...
+    set<string> forwardNotify(s_forwardNotify);
+    for(set<string>::const_iterator j=forwardNotify.begin();j!=forwardNotify.end();++j) {
+      L<<Logger::Warning<<"Relaying notification of domain "<<p->qdomain<<" from "<<p->getRemote()<<" to "<<*j<<endl;
+      Communicator.notify(p->qdomain,*j);
+    }
+  }
+
+  if(::arg().mustDo("slave"))
```
Comment (Habbie) [inline pdns/packethandler.cc]: Is this intentional? What is the behaviour change here?

File: pdns/packethandler.cc (hunk 2 above)
Comment (DrRemorse) [inline pdns/packethandler.cc]: Good call.  Ill make the change and squash tonight.

File: pdns/packethandler.cc (hunk 3 above)
Comment (DrRemorse) [inline pdns/packethandler.cc]: This is intentional to preserve the original pdns behavior due to my conditional change at the top of the function: if(!::arg().mustDo("slave") && ::arg()["forward-notify"].empty()) ...

**Pull request closed without merging**